### PR TITLE
[REVIEW] Add empty column factory

### PR DESCRIPTION
**Pull request by jrhemstad**
Adds a factory for making an empty column. Updates `empty_like` to use this factory. 

Per past discussion, empty compound columns need not contain children.

### CODE REVIEW

File: cpp/tests/copying/utility_tests.cu [hunk 1]
```diff
@@ -87,7 +87,7 @@ void check_empty_string_columns(cudf::column_view lhs, cudf::column_view rhs)
     EXPECT_EQ(lhs.null_count(), 0);
     EXPECT_EQ(lhs.nullable(), false);
     EXPECT_EQ(lhs.has_nulls(), false);
-    EXPECT_EQ(lhs.num_children(), rhs.num_children());
+    // An empty column is not required to have children
```
Comment (trevorsm7) [inline cpp/tests/copying/utility_tests.cu]: Should the same change be made in `expect_column_properties_equal` in  cpp/tests/utilities/column_utilities.cu?

File: cpp/src/copying/copy.cpp [hunk 2]
```diff
@@ -45,14 +46,7 @@ inline mask_state should_allocate_mask(mask_allocation_policy mask_alloc, bool m
  */
 std::unique_ptr<column> empty_like(column_view const& input, cudaStream_t stream)
```
Comment (trevorsm7) [inline cpp/src/copying/copy.cpp]: Would it make sense to remove the unused stream parameter?

Comment (jrhemstad) [reply cpp/src/copying/copy.cpp]: Yep. Done.

File: cpp/tests/copying/utility_tests.cu (hunk 1 above)
Comment (jrhemstad) [reply cpp/tests/copying/utility_tests.cu]: `expect_columns_equal` should check for _exact_ equality. I can't recall who, but someone is adding `expect_columns_equivalent` that checks for semantic equivalence (which should do the check you suggest).

File: cpp/include/cudf/column/column_factories.hpp [hunk 3]
```diff
@@ -23,8 +23,17 @@
...
 #include <rmm/thrust_rmm_allocator.h>
 
-namespace cudf
-{
+namespace cudf {
+
+/**
+ * @brief Creates an empty column of the specified @p type
+ *
+ * An empty column does not contain any elements nor a validity mask.
```
Comment (harrism) [inline cpp/include/cudf/column/column_factories.hpp]:
```suggestion
 * An empty column does not contain any elements or a validity mask.
```

File: cpp/src/copying/copy.cpp [hunk 4]
```diff
@@ -15,108 +15,87 @@
...
-{
-namespace detail
-{
-
-inline mask_state should_allocate_mask(mask_allocation_policy mask_alloc, bool mask_exists) {
+namespace cudf {
+namespace experimental {
+namespace detail {
+namespace {
+inline mask_state should_allocate_mask(mask_allocation_policy mask_alloc,
```
Comment (harrism) [inline cpp/src/copying/copy.cpp]:
```suggestion

inline mask_state should_allocate_mask(mask_allocation_policy mask_alloc,
```

File: cpp/src/copying/copy.cpp [hunk 5]
```diff
@@ -15,108 +15,87 @@
...
+                                       bool mask_exists) {
   if ((mask_alloc == mask_allocation_policy::ALWAYS) ||
       (mask_alloc == mask_allocation_policy::RETAIN && mask_exists)) {
     return UNINITIALIZED;
   } else {
     return UNALLOCATED;
   }
 }
+}  // namespace
```
Comment (harrism) [inline cpp/src/copying/copy.cpp]:
```suggestion

}  // namespace
```

File: cpp/src/copying/copy.cpp (hunk 4 above)
Comment (harrism) [reply cpp/src/copying/copy.cpp]: Should probably have docs for this too.

File: cpp/tests/column/factories_test.cpp [hunk 6]
```diff
@@ -173,4 +174,20 @@ TEST_P(NonNumericFactoryTest, NonNumericThrow) {
...
+TYPED_TEST(EmptyFactoryTest, Empty) {
+  auto type = cudf::data_type{cudf::experimental::type_to_id<TypeParam>()};
+  auto column = cudf::make_empty_column(type);
+  EXPECT_EQ(type, column->type());
+  EXPECT_EQ(column->size(), 0);
+  EXPECT_EQ(0, column->null_count());
+  EXPECT_FALSE(column->nullable());
+  EXPECT_FALSE(column->has_nulls());
+  EXPECT_EQ(0, column->num_children());
+}
```
Comment (harrism) [inline cpp/tests/column/factories_test.cpp]:
```suggestion
}

```

**Pull request merged**